apps/chat/agent/openapi_tool.py

```diff
@@ -0,0 +1,159 @@
+from urllib.parse import parse_qs, urlparse
+
+import httpx
+from celery.utils.text import dedent
+from langchain_core.tools import BaseTool, ToolException
+from pydantic import BaseModel, ConfigDict, Field
+
+from apps.custom_actions.models import CustomAction
+
+
+class OpenAPIToolInput(BaseModel):
+    endpoint: str = Field(..., description="The API endpoint to call")
+    method: str = Field(..., description="HTTP method (GET, POST, PUT, DELETE, etc.)")
+    params: dict | None = Field(..., description="Query parameters for the request")
+    data: dict | None = Field(..., description="Request body data")
+    headers: dict | None = Field(..., description="Custom headers for the request")
+
+
+class OpenAPITool(BaseTool):
+    """A custom tool for making API requests based on OpenAPI specifications. This is a single tool that
+    can handle multiple API endpoints and methods. The tool is configured with a list of CustomAction objects,
+    each of which represents an API service with its own OpenAPI schema. The tool will determine which action
+    to use based on the endpoint and method provided in the input. The tool will then make the API request
+    using the appropriate action and return the response."""
+
+    custom_actions: list[CustomAction]
+
+    name: str = "openapi_request_tool"
+    description: str = dedent(
+        """
+        Makes HTTP requests to API endpoints defined in the OpenAPI specification shown below.
+        Authentication and request formatting are handled based on the OpenAPI schema.
+
+        Generate the full API request details for answering the user question.
+        You should build the API request in order to get a response that is as short as possible, while still
+        getting the necessary information to answer the question. Pay attention to deliberately exclude
+        any unnecessary pieces of data in the API call.
+        """
+    )
+    args_schema: type[OpenAPIToolInput] = OpenAPIToolInput
+    handle_tool_error: bool = True
+    executors: list["ActionExecutor"] = None
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.executors = [ActionExecutor(action) for action in self.custom_actions]
+
+    def _run(self, endpoint: str, method: str, params: dict | None, data: dict | None, headers: dict | None) -> str:
+        for executor in self.executors:
+            if executor.can_run(endpoint, method):
+                return executor.run(endpoint, method, params, data, headers)
```

Comment:
Does the fact that we're returning here mean that only one executer will ever be run (assuming all can actuall run). What if I have multiple `custom_actions` configured?

Reply:
Oh I see now, the LLM will specify the endpoint + method to run. We then loop through all schemata to check if the endpoint exists in it. If it does, we mark that one as runnable. So we basically hit the first `endpoint` with `method` that exists across all specified `custom_actions`.